clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/provider/agent/KubernetesControllersCachingAgent.groovy

```diff
@@ -0,0 +1,448 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.clouddriver.kubernetes.provider.agent
+
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.frigga.Names
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.cats.agent.AgentDataType
+import com.netflix.spinnaker.cats.agent.CacheResult
+import com.netflix.spinnaker.cats.agent.DefaultCacheResult
+import com.netflix.spinnaker.cats.cache.CacheData
+import com.netflix.spinnaker.cats.cache.DefaultCacheData
+import com.netflix.spinnaker.cats.provider.ProviderCache
+import com.netflix.spinnaker.clouddriver.cache.OnDemandAgent
+import com.netflix.spinnaker.clouddriver.cache.OnDemandMetricsSupport
+import com.netflix.spinnaker.clouddriver.kubernetes.deploy.KubernetesUtil
+import com.netflix.spinnaker.clouddriver.kubernetes.model.KubernetesServerGroup
+import com.netflix.spinnaker.clouddriver.kubernetes.provider.view.MutableCacheData
+import com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials
+import com.netflix.spinnaker.clouddriver.kubernetes.cache.Keys
+import groovy.util.logging.Slf4j
+import io.fabric8.kubernetes.api.model.Event
+import io.fabric8.kubernetes.api.model.HorizontalPodAutoscaler
+import io.fabric8.kubernetes.api.model.Pod
+import io.kubernetes.client.Configuration
+import io.kubernetes.client.apis.AppsV1beta1Api
+
+import io.kubernetes.client.ApiClient
+import io.kubernetes.client.models.V1PodList
+import io.kubernetes.client.models.V1beta1StatefulSet
+import io.kubernetes.client.models.V1beta1StatefulSetList
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.AUTHORITATIVE
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.INFORMATIVE
+/**
+ * Created by spinnaker on 9/8/17.
+ */
+@Slf4j
+class KubernetesControllersCachingAgent extends KubernetesCachingAgent implements OnDemandAgent{
+  final OnDemandMetricsSupport metricsSupport
+
+  //final ApiClient newClient
+  static final String DEPLOYMENT_ANNOTATION = "deployment.kubernetes.io"
+  final String category = 'serverGroup'
+  static final Set<AgentDataType> types = Collections.unmodifiableSet([
+    AUTHORITATIVE.forType(Keys.Namespace.APPLICATIONS.ns),
+    AUTHORITATIVE.forType(Keys.Namespace.CLUSTERS.ns),
+    INFORMATIVE.forType(Keys.Namespace.LOAD_BALANCERS.ns),
+    AUTHORITATIVE.forType(Keys.Namespace.SERVER_GROUPS.ns),
```

Comment:
I don't think these should live in the SERVER_GROUP namespace because now you'll have two caching agents producing keys in the same "shard".